kiwmi/desktop/output.c

```diff
@@ -381,8 +420,126 @@ new_output_notify(struct wl_listener *listener, void *data)
     wl_signal_emit(&desktop->events.new_output, output);
 }
 
+static void
+output_manager_configure(
+    struct kiwmi_server *server,
+    struct wlr_output_configuration_v1 *configuration,
+    bool test)
+{
+    struct wlr_output_configuration_head_v1 *head;
+    struct wlr_output_configuration_head_v1 *bad_head = NULL;
+    wl_list_for_each (head, &configuration->heads, link) {
+        struct wlr_output *wlr_output = head->state.output;
+
+        wlr_output_enable(wlr_output, head->state.enabled);
+        if (head->state.enabled) {
+            if (head->state.mode) {
+                wlr_output_set_mode(wlr_output, head->state.mode);
+            } else {
+                wlr_output_set_custom_mode(
+                    wlr_output,
+                    head->state.custom_mode.width,
+                    head->state.custom_mode.height,
+                    head->state.custom_mode.refresh);
+            }
+
+            wlr_output_set_transform(wlr_output, head->state.transform);
+            wlr_output_set_scale(wlr_output, head->state.scale);
+        }
+
+        if (!wlr_output_test(wlr_output)) {
+            bad_head = head;
+            break;
+        }
+    }
+
+    wl_list_for_each (head, &configuration->heads, link) {
+        if (bad_head != NULL || test) {
+            wlr_output_rollback(head->state.output);
+            if (head == bad_head) {
+                break;
+            } else {
+                continue;
+            }
+        }
+        struct kiwmi_output *output = head->state.output->data;
+        bool enable_changed         = WLR_OUTPUT_STATE_ENABLED
+            == (output->wlr_output->pending.committed
+                & WLR_OUTPUT_STATE_ENABLED);
+        if (head->state.enabled) {
+            if (enable_changed) {
+                wlr_output_layout_add(
+                    server->desktop.output_layout,
+                    output->wlr_output,
+                    head->state.x,
+                    head->state.y);
+                wl_signal_emit(&output->desktop->events.new_output, output);
+            } else {
+                bool moved = output->wlr_output->pending.committed
+                    & (WLR_OUTPUT_STATE_MODE | WLR_OUTPUT_STATE_SCALE
+                       | WLR_OUTPUT_STATE_TRANSFORM);
+                struct wlr_output_layout_output *layout_output =
+                    wlr_output_layout_get(
+                        server->desktop.output_layout, output->wlr_output);
+                moved |= head->state.x != layout_output->x
+                    || head->state.y != layout_output->y;
```

Comment:
I wouldn’t be scared of making this two variables, `moved` & `resized`. Would help code understanding and also make it easier to add a separate `move` event to `kiwmi_output` if desired.

Reply:
Fair, though I'd recommend we keep moves and resizes in the same event, seeing as the vast majority of configs will be doing exactly the same thing in each case.